core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java

```diff
@@ -45,6 +53,14 @@
     private final PlannerContext plannerContext;
     private final StatementAnalyzerFactory statementAnalyzerFactory;
 
+    private final NonEvictableCache<NodeRef<Expression>, Map<NodeRef<Expression>, Type>> expressionTypesCache = buildNonEvictableCache(
```

Comment:
preserve `ExpressionAnalyzer` instead (it caches types internally). It will also work with nested expressions compared to `expressionTypesCache` here